storm-core/src/jvm/org/apache/storm/daemon/supervisor/ContainerLauncher.java

```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Launches containers
+ */
+public abstract class ContainerLauncher {
+    private static final Logger LOG = LoggerFactory.getLogger(ContainerLauncher.class);
+    
+    /**
+     * Factory to create the right container launcher 
+     * for the config and the environment.
+     * @param conf the config
+     * @param supervisorId the ID of the supervisor
+     * @param sharedContext Used in local mode to let workers talk together without netty
+     * @return the proper container launcher
+     * @throws IOException on any error
+     */
+    public static ContainerLauncher mk(Map<String, Object> conf, String supervisorId, IContext sharedContext) throws IOException {
+        if (ConfigUtils.isLocalMode(conf)) {
+            return new LocalContainerLauncher(conf, supervisorId, sharedContext);
+        }
+        
+        ResourceIsolationInterface resourceIsolationManager = null;
+        if (Utils.getBoolean(conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN_ENABLE), false)) {
+            resourceIsolationManager = Utils.newInstance((String) conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN));
+            resourceIsolationManager.prepare(conf);
+            LOG.info("Using resource isolation plugin {} {}", conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN), resourceIsolationManager);
+        }
+
+        if (Utils.getBoolean(conf.get(Config.SUPERVISOR_RUN_WORKER_AS_USER), false)) {
+            return new RunAsUserContainerLauncher(conf, supervisorId, resourceIsolationManager);
+        }
+        return new BasicContainerLauncher(conf, supervisorId, resourceIsolationManager);
+    }
+    
+    protected ContainerLauncher() {
+        //Empty
+    }
+
+    /**
+     * Launch a container in a given slot
+     * @param port the port to run this on
+     * @param assignment what to launch
+     * @param state the current state of the supervisor
+     * @return The container that can be used to manager the processes.
+     * @throws IOException on any error 
+     */
+    public abstract Container launchContainer(int port, LocalAssignment assignment, LocalState state) throws IOException;
+    
+    /**
+     * Recover a container for a running process
+     * @param port the port the assignment is running on
+     * @param assignment the assignment that was launched
+     * @param state the current state of the supervisor
+     * @return The container that can be used to manage the processes.
+     * @throws IOException on any error
+     */
+    public abstract Container recoverContainer(int port, LocalAssignment assignment, LocalState state) throws IOException;
+    
+    /**
+     * Try to recover a container with just the worker ID.  
+     * The result is really only useful for killing the container.
+     * @param workerId the id of the worker to use
+     * @return a Container that can be killed.
+     * @throws IOException on any error
+     */
+    public abstract Killable recoverContainer(String workerId) throws IOException;
```

Comment:
I like the idea of using an interface specific to the scenario. Only comment is we should update the javadoc to say it returns a Killable instead of a Container.